### PR TITLE
CLI: Rework ModemManager check and add udev check

### PR DESCRIPTION
**Pull request by Erovia**
## Description
Check the udev rules on the system for the recommended rules. Print warning if not found.
If the udev rules for catalina bootloader are not present, check if ModemManager is running, because that could cause troubles with flashing. In this scenario the `doctor` strongly advices fixing the problem.

**Note**: I tested it the best I could, but most likely need some work before merging. I'm opening this PR in order to get early feedback and help with testing.

## Types of Changes

- [x] Core
- [ ] Bugfix
- [x] New feature
- [x] Enhancement/optimization
- [ ] Keyboard (addition or update)
- [ ] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Issues Fixed or Closed by This PR

* 

## Checklist

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

File: lib/python/qmk/cli/doctor.py [hunk 1]
```diff
@@ -49,23 +49,43 @@ def doctor(cli):
...
+        udev_dir = "/etc/udev/rules.d/"
+        # These are the recommended udev rules
+        desired_rules = {"dfu": {'SUBSYSTEMS=="usb", ATTRS{idVendor}=="03eb", ATTRS{idProduct}=="2ff4", MODE:="0666"',
+                                 'SUBSYSTEMS=="usb", ATTRS{idVendor}=="03eb", ATTRS{idProduct}=="2ffb", MODE:="0666"',
+                                 'SUBSYSTEMS=="usb", ATTRS{idVendor}=="03eb", ATTRS{idProduct}=="2ff0", MODE:="0666"'},
+                         "tmk": {'SUBSYSTEMS=="usb", ATTRS{idVendor}=="feed", MODE:="0666"'},
+                         "input-club": {'SUBSYSTEMS=="usb", ATTRS{idVendor}=="1c11", MODE:="0666"'},
+                         "stm32": {'SUBSYSTEMS=="usb", ATTRS{idVendor}=="1eaf", ATTRS{idProduct}=="0003", MODE:="0666"',
+                                   'SUBSYSTEMS=="usb", ATTRS{idVendor}=="0483", ATTRS{idProduct}=="df11", MODE:="0666"'},
+                         "catalina": {'ATTRS{idVendor}=="2a03", ENV{ID_MM_DEVICE_IGNORE}="1"',
```
Comment (fauxpark) [inline lib/python/qmk/cli/doctor.py]:
```suggestion
                         "caterina": {'ATTRS{idVendor}=="2a03", ENV{ID_MM_DEVICE_IGNORE}="1"',
```

File: lib/python/qmk/cli/doctor.py [hunk 2]
```diff
@@ -49,23 +49,43 @@ def doctor(cli):
...
+            # Check if the desired rules are among the currently present rules
+            for bootloader, rules in desired_rules.items():
+                if not rules.issubset(current_rules):
+                    # If the rules for catalina are not present, check if ModemManager is running
+                    if bootloader == "catalina":
+                        if shutil.which("systemctl"):
+                            mm_check = subprocess.run(["systemctl", "--quiet", "is-active", "ModemManager.service"], timeout=10)
+                            if mm_check.returncode == 0:
+                                ok = False
+                                cli.log.warn("{bg_yellow}Detected ModemManager without udev rules. Please either disable it or set the appropriate udev rules if you are using a Pro-Micro.")
```
Comment (fauxpark) [inline lib/python/qmk/cli/doctor.py]:
```suggestion
                                cli.log.warn("{bg_yellow}Detected ModemManager without udev rules. Please either disable it or set the appropriate udev rules if you are using a Pro Micro.")
```

File: lib/python/qmk/cli/doctor.py (hunk 1 above)
Comment (Erovia) [reply lib/python/qmk/cli/doctor.py]: Right, [this line in the docs](https://github.com/qmk/qmk_firmware/blame/master/docs/faq_build.md#L50) throw me off balance.

Comment (fauxpark) [reply lib/python/qmk/cli/doctor.py]: Yeah, that should really be changed too 😄

**Comment by Erovia** [comment]
Rebased on master.

**Pull request merged**